File: 04-1.rb
#!/usr/bin/env ruby

$haystack = File.read('04.input').lines.map(&:strip).map(&:chars)

def find_dir(needle, x, y, dx, dy)
  return true if needle.empty?
  return false if x < 0 or
                  x >= $haystack[0].size or
                  y < 0 or
                  y >= $haystack.size
  return false unless $haystack[y][x] == needle[0]

  find_dir(needle[1..], x + dx, y + dy, dx, dy)
end

def find(x, y)
  needle = 'XMAS'

  count = 0
  [-1, 0, 1].each do |dx|
    [-1, 0, 1].each do |dy|
      next if dx.zero? and dy.zero?

      count += 1 if find_dir(needle, x, y, dx, dy)
    end
  end
  count
end

count = 0
$haystack.each_index do |y|
  $haystack[0].each_index do |x|
    count += find(x, y)
  end
end

puts count

File: 04-1.py
#!/usr/bin/env python

haystack = [row.strip() for row in open('04.input').readlines()]

def find_dir(needle, x, y, dx, dy):
    if len(needle) == 0:
        return True
    if x < 0 or x >= len(haystack[0]) or \
       y < 0 or y >= len(haystack):
        return False
    if haystack[y][x] != needle[0]:
        return False

    return find_dir(needle[1:], x + dx, y + dy, dx, dy)

def find(x, y):
    needle = 'XMAS'

    count = 0
    for dx in [-1, 0, 1]:
        for dy in [-1, 0, 1]:
            if dx == 0 and dy == 0:
                continue

            if find_dir(needle, x, y, dx, dy):
                count += 1

    return count


count = 0
for y in range(len(haystack)):
    for x in range(len(haystack[0])):
        count += find(x, y)

print(count)
